Initial Version of Hathi Reader

mwbenowitz opened the pull request:
This lambda function reads from CSV files formatted according to the standard output from HathiTrust. At present it will only handle local files, but contains a stub method that will allow it to harvest daily updates from the HathiTrust repository.

The function parses each row from the CSV files into an SFR-compliant data model. This includes an update to that data model in a new class for rights information. This allows a more robust (and extensible) model for rights metadata that will allow us to track the exact rights status for the various versions of the materials we are ingesting.

mwbenowitz committed:
Initial Commit of HathiTrust reader

This includes the core code that parses data from a local source
into the SFR data model

mwbenowitz committed:
Initial Commit of HathiTrust reader

This includes the core code that parses data from a local source
into the SFR data model

mwbenowitz committed:
Fix typo in requirements.txt file

mwbenowitz committed:
Fix typo in requirements.txt file

mwbenowitz committed:
Remove extra boilerplate script in travis config

mwbenowitz committed:
Remove extra boilerplate script in travis config

mwbenowitz committed:
Merge branch 'development' of https://github.com/NYPL/sfr-hathitrust-reader into development

mwbenowitz committed:
Add missing lxml to requirements.txt

mwbenowitz committed:
SFR-261 Fix minor glitches in hathi record ingest

Two minor issues cropped up during development. First,
the output kinesis stream should not be hard coded, but
set in an environment var. Second, the output rights and
identifiers for Hathi records should be a) there and b) always
encoded as lists so that they can be dynamically handled by the
database manager (There should never be an instance in our data
model where there is specified one-to-one relationships, at least
not yet)

mwbenowitz committed:
SFR-260 Update test cases

seanredmond commented on lib/countryParser.py:
```diff
@@ -0,0 +1,49 @@
+    with countryXML as countryXML:
+        countryTree = etree.parse(countryXML)
+        codeList = countryTree.getroot()
+        countries = codeList.find('{info:lc/xmlns/codelist-v1}countries')
```
Just a style thing, but more pythonic would be to construct the `dict` out of a list comprehension:

    countryTrans = dict([(x.find('{info:lc/xmlns/codelist-v1}code').text, x.find('{info:lc/xmlns/codelist-v1}name').text) for x in codeList.findall('.//{info:lc/xmlns/codelist-v1}country')])

mwbenowitz replied:
Very true, looking at this again I don't know why I was looping to get the child elements of each country, so inefficient!

seanredmond commented on service.py:
```diff
@@ -0,0 +1,263 @@
+    # retrieve specific values
+    hathiDict = dict(zip(columns, row))
+    # Generate a hathi record object with the source dict
+    hathiRec = HathiRecord(hathiDict)
```
Would you instantiate a `HathiRecord` without building it? My first reaction is that `buildWork()`, `buildInstance()` etc. could just be part of construction. But maybe you want the option of logging the steps (and you don't want to build that into the `HathiRecord` object)?   

I'm not sure what the trade-offs are. There's kind of a mix of encapsulation (like `buildWork()`) and transparency (like directly manipulating `hathiRec.work.instances`). I wonder if it could be clearer?

mwbenowitz replied:
I think it can be and personally I prefer to keep most of this logic within the class itself/outside of the general handler (division of responsibility and all that). I added a new function `buildDataModel` that better encapsulates the logic here and which feels more clear to me. It is a little less transparent but breaks it down it more logical level of details so someone reading the code can "descend" to each level as necessary to get more detail.

You could certainly create an empty `HathiRecord` object and then build the source dict, perhaps from multiple sources (I'm thinking if we perhaps end up using the API to get more detailed info from Hathi on top of what's available in the CSV files).

seanredmond commented on service.py:
```diff
@@ -0,0 +1,263 @@
+            'status': 200,
+            'type': 'work',
+            'method': 'insert',
+            'data': hathiRec.work
```
Sorry if this is a dumb question. The work is put into Kinesis -- does anything happen with the instance or item?

mwbenowitz replied:
The instance and item are added as children record to the work as part of the `parseRow` method, so they are included here.

seanredmond commented on lib/hathiRecord.py:
```diff
@@ -0,0 +1,590 @@
+        self.parsePubPlace(self.ingest['pub_place'], countryCodes)
+
+        # Instances and works will share many identifiers
+        instanceIdentifiers = [
```
Should this be placed at the top of the file where some other lists are constructed? The identical list seems to be used in `buildWork()`

mwbenowitz replied:
Yes, I was anticipating these being different, but they're not.

seanredmond commented on lib/countryParser.py:
```diff
@@ -0,0 +1,49 @@
+logger = createLog('country_code_generator')
+
+
+def loadCountryCodes():
```
Might be worth just turning this list of country codes into a module.

mwbenowitz replied:
I'm not sure if this will end up being reused (most places use the ISO country codes, not the MARC ones), but if it is then I probably will turn it into a module (especially since one doesn't seem to exist at all)

mwbenowitz committed:
SFR-260 List comprehension for country name generator

mwbenowitz committed:
SFR-260 HathiRecord builder refactor

On receiving some readability/understandability feedback this
commit refactors the record building logic into a new method
with the HathiRecord class. This makes the flow of the function
smooth and encapsulates the work being done in a more logical and
hierichical manner. It should be easier to see both the high
level operation and the detailed methods depending on your interest.

mwbenowitz commented:
Pushed a few commits to make the function more readable/understandable. These do not change the logic of the function at all, but hopefully make it more maintainable in the long run

mwbenowitz committed:
Add TODOs and brief explanation of more complex metadata
